hudi-client/hudi-client-common/src/main/java/org/apache/hudi/index/HoodieIndexUtils.java

```diff
@@ -243,6 +247,56 @@ private static <R> HoodieData<HoodieRecord<R>> getExistingRecords(
         .getMergedRecords().iterator());
   }
 
+  /**
+   * getExistingRecords will create records with expression payload so we overwrite the config.
+   * Additionally, we don't want to restore this value because the write will fail later on.
+   * We also need the keygenerator so we can figure out the partition path after expression payload
+   * evaluates the merge.
+   */
+  private static BaseKeyGenerator maybeGetKeygenAndUpdatePayload(HoodieWriteConfig config) {
+    if (config.getPayloadClass().equals("org.apache.spark.sql.hudi.command.payload.ExpressionPayload")) {
+      config.setValue(HoodiePayloadConfig.PAYLOAD_CLASS_NAME.key(), HoodiePayloadConfig.PAYLOAD_CLASS_NAME.defaultValue());
```

Comment:
what incase we switch the default payload class later? should we instead hardcode the payload class.